### PR TITLE
initial commit of Set-DbaMaxDop

### PR DESCRIPTION
**Pull request by ClaudioESSilva**

### CODE REVIEW

**Comment by dansqldba** [comment]
This PR includes the Test-DbaMaxDop as a changed file, can you remove the file and commit again please?

**Comment by ClaudioESSilva** [comment]
Done @dansqldba

**Comment by ClaudioESSilva** [comment]
Set of tests for this PR
SQL Server 2000
- [ ] Don't run on version 2000 and throw an message

SQL Server 2005 - 2016
- [ ] Run command on a single instance. Do not specify the -MaxDop parameter. Must change the MaxDop for the recommended one.
- [ ] Run command on a single instance. Specify the -MaxDop parameter to other than the current value. Must change the MaxDop for the specified one.
- [ ] Run command on a single instance. Specify the -Databases parameter and select one. Must give warning saying "Server '$servername' does not supports Max DOP configuration per database. Run the command again without -Databases parameter. Skipping."
- [ ] Run command on mutiple instances. Do not specify the -MaxDop parameter. Must change the MaxDop for the recommended one.
- [ ] Run command on mutiple instances. Specify the -MaxDop parameter to other than the current value. Must change the MaxDop for the specified one.
- [ ] Run command on mutiple instance (one of them 2016) and specify -Database. Should see a warning for non-2016 version saying "Server '$servername' does not supports Max DOP configuration per database. Run the command again without -Databases parameter. Skipping.". On 2016 instance the database should be configured with recommended value.

SQL Server 2005 - 2014 ONLY
- [ ] Run command on a single instance and specify -Database. Should see a warning "Server '$servername' does not supports Max DOP configuration per database. Run the command again without -Databases parameter. Skipping."

SQL Server 2016
- [ ] Run command on a single instance and specify -Database. If using -Verbose should see a output "Server '$servername' supports Max DOP configuration per database."
- [ ] Run command on a single instance and specify -Database. Do not specify the -MaxDop parameter. The configuration must be applied only on the Database(s) defined.
- [ ] Preparation. Set the same MaxDop for every database (User databases). Set MaxDop on instance level to 0. Run the command on that instance. The result shoulb be all databases max dop set to 0 and instance set to the databases max dop. That way if we add new databases as the default is 0, all databases are inherith the instance level configuration.

**Comment by ctrlbold** [comment]
Thorough AF 😍

**Comment by ctrlbold** [comment]
Any takers? 

- [ ] Working/useful help content, including link to command on dbatools web site
- [ ] All examples work as advertised
- [ ] Does not contain template content
- [ ] Does not contain excessive/unnecessary amounts of comments
- [ ] Works remotely
- [ ] Works locally
- [ ] Works on lower versions or throws error specifying version not supported
- [ ] Works with named instances
- [ ] Works with clustered instances
- [ ] Handles offline/read only databases
- [ ] Supports multiple servers (at the command line or piped from Get-SqlRegisteredServerName)
- [ ] No un-handled errors which stop the command working with multiple servers

**Comment by ctrlbold** [comment]
- [ ] Does not support "sql2005","sql2008" | Set-DbaMaxDop
- [ ] Set-DbaMaxDop -SqlServer sql2005 -MaxDop 1 -Databases mydb should throw warning but it does not.
- [ ] Set-DbaMaxDop -SqlServer sql2005,sql2008, sql2014 -MaxDop 10 -Data does not autopopulate - I'm assuming $sqlserver[0] was not used in the autopopulation thing. But should it? I kinda like that it doesn't.
- [ ] I would expect for database DOP to change *only* if I specified -Databases, so Set-DbaMaxDop -SqlServer sql2016 -Verbose had two issues - 
1. it changed all of my database dops (did it? Maybe it's just reporting for each one and it didn't change the database, just the instance over and over)
2. the output was wrong

image

If it is changing per db without dbs set, please only change instance level if databases is not set since it's inherited. 

- [ ] Set-DbaMaxDop -SqlServer sql2016 -Databases db1 -MaxDop 55

    Instance                   : sql2016
    Database                   : db1
    OldDatabaseMaxDopValue     : 0
    CurrentDatabaseMaxDopValue : 55
    OldInstanceMaxDopValue     : 5
    CurrentInstanceMaxDop      : 5

This made me worried that I had changed the server dop when I only intended to change the database dop. So I tested with -MaxDop 10 and saw it only changed the database. 

Can  OldInstanceMaxDopValue be dropped and CurrentInstanceMaxDop  changed to InstanceMaxDop when -Databases is specified?

- [ ] Set db3 to offline on sql2016 then ran Set-DbaMaxDop -SqlServer sql2016 -Databases db3  -- it just returned nothing

**Comment by ClaudioESSilva** [comment]
Changes made. Please re-test.
The only one will be for a future enhancement is the "Set db3 to offline on sql2016 then ran Set-DbaMaxDop -SqlServer sql2016 -Databases db3 -- it just returned nothing"

File: functions/Set-DbaMaxDop.ps1 [hunk 1]
```diff
@@ -0,0 +1,295 @@
...
+{
+<# 
+.SYNOPSIS 
+Sets SQL Server max dop then displays information relating to SQL Server Max DOP configuration settings. Works on SQL Server 2005-2016.
+
+.DESCRIPTION 
+Uses the Test-DbaMaxDop command to get the recommended value if -MaxDop parameter is not specified.
+
+These are just general recommendations for SQL Server and are a good starting point for setting the "max degree of parallelism" option.
+You can set MaxDop databsae scoped configurations if the server is an 2016.
```
Comment (ctrlbold) [inline functions/Set-DbaMaxDop.ps1]: database is spellled wrong and should be version 2016

**Pull request merged**